include/boost/geometry/extensions/algorithms/dissolve_using_correct.hpp

```diff
@@ -0,0 +1,533 @@
+		ring.push_back(ring.front());
+
+}
+
+template <typename Ring, typename PseudoVertices, typename Area>
+inline std::vector<std::pair<Ring, Area>> dissolve_generate_rings(
+			PseudoVertices &pseudo_vertices,
+    		std::set<pseudo_vertice_key, compare_pseudo_vertice_key> const &all_start_keys, 
+			geometry::order_selector order, Area remove_spike_min_area)
+{
+	std::vector<std::pair<Ring, Area>> result;
+
+	// Generate all polygons by tracing all the intersections
+	// Perform union to combine all polygons into single polygon again
+	auto start_keys = all_start_keys;
+    while(!start_keys.empty()) {    
+		Ring new_ring;
+
+		// Store point in generated polygon
+		auto push_point = [&new_ring](auto const &p) { 
+            if(new_ring.empty() || geometry::comparable_distance(new_ring.back(), p) > 0) {
+                new_ring.push_back(p);
+			}
+		};
+
+		// Store newly generated ring
+		auto push_ring = [&result, remove_spike_min_area](auto& new_ring) {
+			auto const area = geometry::area(new_ring);
+			if(std::abs(area) > remove_spike_min_area) {
+		    	result.push_back(std::make_pair(std::move(new_ring), area));
+			}
+		};
+
+        auto i = pseudo_vertices.find(*start_keys.begin());	
+		
+		using point_t = geometry::point_type_t<Ring>;
+    
+    	std::vector< std::pair<point_t, std::size_t> > start_points;
+		start_points.push_back(std::make_pair(i->second.p, 0));
+
+		// Check if the outer or inner ring is closed
+		auto is_closed = [&new_ring, &start_points, &push_ring](auto const& p) {
+			for(auto const &i: start_points) {
+				if(new_ring.size() > i.second+1 && geometry::comparable_distance(i.first, p) == 0) {
+					if(i.second == 0) return true;
+
+					// Copy the new inner ring
+					Ring inner_ring(new_ring.begin() + i.second, new_ring.end());
+					push_ring(inner_ring);
+
+					// Remove the inner ring
+					new_ring.erase(new_ring.begin() + i.second, new_ring.end());
+				}
+			}
+			return false;
+		};
+
+        do {
+            auto const &key = i->first;
+            auto const &value = i->second;
+        
+			// Store the point in output polygon
+			push_point(value.p);
+            
+			// Remove the key from the starting keys list
+			auto compare_key = [&key](pseudo_vertice_key const &i) {
+				return (key.index_1 == i.index_1 && key.index_2 == i.index_2 && key.scale == i.scale && key.reroute == i.reroute);
+			};
+
+			start_keys.erase(key);
+
+			// Store possible new inner ring starting point
+			if(all_start_keys.find(key) != all_start_keys.end())
+				start_points.push_back(std::make_pair(value.p, new_ring.size() - 1));
+
+            if(key.reroute) {
+				// Follow by-pass
+                i = pseudo_vertices.find(value.link);
+			} else {
+				// Continu following original polygon
+                ++i;
+                if(i == pseudo_vertices.end())
+                    i = pseudo_vertices.begin();
+            }
+
+			// Repeat until back at starting point
```

Comment:
I believe this `while` can lead to infinite loops. I had it once when changing a thingy.
But in the current unit test, it is always fine.